[New Concept]: Range (#1569)
exercism/ruby: concepts/ranges/about.md

# Ranges

[Ranges][ranges] represents an interval between two values.
The most common types that support ranges are `Int` and `String`.
They can be used for many things like quickly creating a collection, slicing strings, checking if a value is in a range and iteration.
They are created using the range operator `..` or `...` (inclusive and exclusive respectively).

```ruby
1..5 # => 1, 2, 3, 4, 5
1...5 # => 1, 2, 3, 4

> ```

Reviewer: The return is not what this statement returns, as we can see.  Using IRB, since IRB "evaluated" syntax is being used here in the code example:

~~~suggestion

In IRB:
```ruby
>> (1..5)
=> 1..5
>> (1...5)
=> 1...5
```
~~~

What do you think?  The return is not like it would be with an Array.

Probably not in this section, but we can coerce this to an Array to show what is originally here:

```irb
>> (1..5).to_a
=> [1, 2, 3, 4, 5]
>> (1...5).to_a
=> [1, 2, 3, 4]
```

The reason for having two range operators is to allow to create ranges that are inclusive or exclusive of the end value, which can be useful when for example working with indexes, that are zero based.

> Ranges can also be created using the `Range` initializer.
> 
> ```ruby
> Range.new(1, 5) # => 1, 2, 3, 4, 5
> ```

Reviewer: ```suggestion
Ranges can also be created using the `Range` initializer.

Exploring in IRB:

```ruby
>> Range.new(1, 5)
=> 1..5
```
~~~

````exercism/note
When creating a range in Ruby using the range operators `..` or `...`, and wanting to call a method on the range, you need to wrap the range in parentheses.
This is because the otherwise will the method be called on the 2nd argument of the range operator.

```ruby
(1..5).sum # => 15
1..5.sum # => Error: undefined method `sum' for 5:Integer (NoMethodError)
```
````

## Getting substrings

When wanting to slice a string, you can use the range operator to get a substring.

> That is by creating a range with the start and end index of the substring.

Reviewer: ```suggestion
That is, by creating a range with the start and end index of the sub-string.
```

```ruby
"Hello World"[0..4] # => "Hello"
"Hello World"[6..10] # => "World"
```

You can also use negative indexes to get the substring from the end of the string.

```ruby
"Hello World"[-5..-1] # => "World"
"Hello World"[6..-4] # => "Wor"
```

## Range methods

Ranges do have a set of methods that can be used to work with them.
These methods when for example want to get the sum of all the values in the range, or check if the range includes a value.

> | Method                  | Description                                                             | Example                         |
> | ----------------------- | ----------------------------------------------------------------------- | ------------------------------- |
> | [`sum`][sum]            | Returns the sum of all the values in the range                          | `(1..5).sum # => 15`            |
> | [`size`][size]          | Returns the size of the range                                           | `(1..5).size # => 5`            |
> | [`include?`][indlude] | Returns `true` if the range includes the given value, otherwise `false` | `(1..5).include?(3) # => true` |

Reviewer: ```suggestion
| Method                  | Description                                                             | Example                         |
| ----------------------- | ----------------------------------------------------------------------- | ------------------------------- |
| [`sum`][sum]            | Returns the sum of all the values in the range                          | `(1..5).sum # => 15`            |
| [`size`][size]          | Returns the size of the range                                           | `(1..5).size # => 5`            |
| [`include?`][indlude]   | Returns `true` if the range includes the given value, otherwise `false` | `(1..5).include?(3) # => true` |
```

## Endless & Beginless ranges

> There are two special types of ranges, the endless and beginless ranges.

Reviewer: I wonder if they are really special ranges, or if this is more "syntactic sugar" that lets us express the normal ranges like this.

> This means that the range has no beginning or end.

Reviewer: If we terminate them with negative infinity and positive infinity, then we can see that they do have a beginning or end.

> The endless or beginless range has there start or end value being `nil`, but when defining the range so can nil be omitted.

Reviewer: ```irb
irb(main):004:0> Range.new(nil, 5)
ArgumentError: bad value for range
        from (irb):4:in `initialize'
        from (irb):4:in `new'
        from (irb):4
        from :0
irb(main):005:0> (nil..5)
ArgumentError: bad value for range
        from (irb):5
        from :0
irb(main):006:0> 
```

I do not think this is true, demonstrably.

Author: ![image](https://github.com/exercism/ruby/assets/69751659/e1f50a2b-a20f-46ec-8185-221e3958e7d9)

Looking in the docs are there also mention of nil: https://rubyapi.org/3.2/o/range.

Reviewer: Thanks.  That is why I said _think_... and did it in an REPL, to verify.  Seems like it has changed for newer versions of Ruby.

So I would then expect it to no longer be a terminated by infinity anything, and perhaps "undefined" in behavior.  I will definitely look at the source code.  Pretty sure the use will not change from `1..infinity` as `1..nil` though.  (That would be surprising, but `nil` evaluates as 0 if we do `to_i`  and as `0.0` if we do `nil.to_f` so definitely worth investigating.

Reviewer: Oh, you sent a picture too.  Whatever it is, I a sure it supports your statement.  Not at a good device at the moment to get interpretations for pictures.

Reviewer: Just for information, I happened to have a Pry session open, where Pry has a bug that does not show this as it is supposed to.  That was the reason it was broken as shown.  Not the syntax itself. "Bad Pry!"

> Using beginless and endless ranges is useful when you want to for example slice a string from the beginning or to the end.

Reviewer: ```suggestion
Using beginless and endless ranges is useful when you want to, for example, slice a string from the beginning or to the end.
```

```ruby
"Hello World"[0..] # => "Hello World"
"Hello World"[4..] # => "o World"
"Hello World"[..5] # => "Hello"
```

```exercism/caution

> If not used on a collection, the endless range can cause an infinite loop, if not used with caution.

Reviewer: ```suggestion
If not used on a collection, the endless range can cause an endless sequence, if not used with caution.
```

Because it does not loop, it is more an endless sequence than and endless loop.

Author: Yeah, I mean it could cause an infinte loop if you iterate over the sequence, but yes I think saying it is an endless sequence is better.

```

## String ranges

Strings can also be used in ranges and allow one to get an interval of strings between two strings.

> But its behavior is a bit different than with chars, when using multiple characters in a string range.

Reviewer: ```suggestion
But its behavior is a bit different than with `chars`, when using multiple characters in a string range.
```

I believe you may be referring to the method, and so that should be code formatted.

Its behavior can become when doing more complex string ranges, so use it with caution.

```ruby

> "aa".."az" # => "aa", "ab", "ac", ..., "az"

Reviewer: ```suggestion
In IRB:

>> "aa".."az"
=> "aa", "ab", "ac", # continues until  "az"
```

We could use a smaller sequence as well, and still demonstrate this, so we do not have to allude to the missing content.

"Hello World"[0..4] # => "Hello"
"Hello World"[6..10] # => "World"
```

You can also use negative indexes to get the substring from the end of the string.

```ruby
"Hello World"[-5..-1] # => "World"
"Hello World"[6..-4] # => "Wor"
```

## Range methods

Ranges do have a set of methods that can be used to work with them.
These methods when for example want to get the sum of all the values in the range, or check if the range includes a value.

| Method                  | Description                                                             | Example                         |
| ----------------------- | ----------------------------------------------------------------------- | ------------------------------- |
| [`sum`][sum]            | Returns the sum of all the values in the range                          | `(1..5).sum # => 15`            |
| [`size`][size]          | Returns the size of the range                                           | `(1..5).size # => 5`            |
| [`include?`][indlude]   | Returns `true` if the range includes the given value, otherwise `false` | `(1..5).include?(3) # => true` |

## Endless & Beginless ranges

A range can be endless and beginless.

> The endless or beginless range has there start or end value being `nil`, but when defining the range so can nil be omitted.

Reviewer: ```suggestion
The endless or beginless range has theirr start or end value being `nil`, but when defining the range so can nil be omitted.
```

Or 

```suggestion
The endless or beginless range has its start or end value being `nil`, but when defining the range so can nil be omitted.
```

# Ranges

[Ranges][ranges] represents an interval between two values.
The most common types that support ranges are `Int` and `String`.
They can be used for many things like quickly creating a collection, slicing strings, checking if a value is in a range and iteration.
They are created using the range operator `..` or `...` (inclusive and exclusive respectively).

```ruby
1..5  # => 1..5
1...5 # => 1...5

(1..5).to_a # => [1, 2, 3, 4, 5]
(1...5).to_a # => [1, 2, 3, 4]
```

The reason for having two range operators is to allow to create ranges that are inclusive or exclusive of the end value, which can be useful when for example working with indexes, that are zero based.

Ranges can also be created using the `Range` initializer.

```ruby

> Range.new(1, 5) # => 1, 2, 3, 4, 5

Reviewer 2: ```suggestion
Range.new(1, 5) # A range containing 1, 2, 3, 4, 5
```
